### PR TITLE
Introducing `filterRanges` Function and Discussion on a possible `filterSimilarities`

### PR DESCRIPTION
**Pull request by philouail**
Hi, I wanted to suggest the implementation of this function `filterRanges`, which allows the user to filter a `spectra` object based on any variables of its `spectraData` suing a range of values that the user inputs . I believe this would be great for multiple reasons:

- Allow multiple filtering to happen at the same time.
- Allow for seamless integration of filtering based on any future-added or user-specific spectraData variable.
- Prevent the creation of many variable-specific filtering functions in the future.

Welcoming any feedback on the code and PR :)

On the other hand, I would also like to suggest implementing a similar function such as `filterSimilarities` (not sure about the name), which would be implemented in a similar way for filtering based on *similarity* to specific values for variables within the `spectraData` of the object rather than within a range. Thus, it would have the same advantages as the `filterRanges` function. It would prevent variable-specific functions and make the code easier to combine multiple filters.

I can however see the counter argument that this might not be as user-friendly as variable-specific functions. I would love to get your opinion @jorainer and @lgatto on this. If you are interested i can come up with a more detailed proposition of implementation.

### CODE REVIEW

**Review by jorainer — CHANGES_REQUESTED** [review]
Thank you @philouail ! Great PR! Some change requests and questions.

Also, something strange happened with the roxygen documentation.

**Comment by jorainer** [comment]
To me this function is great and also the `filterSimilarities` - that I would rather suggest to call `filterValues` would make a lot of sense to me. @lgatto what's your opinion on that?

We have already a use case in which we want to extract/filter spectra with their precursor m/z and retention time within respective ranges (from MS1 data).

File: R/Spectra.R [hunk 1]
```diff
@@ -2396,6 +2413,44 @@ setMethod("reset", "Spectra", function(object, ...) {
...
+              }
+              query <- spectraData(object)[, spectraVariables]
+              nc <- ncol(query)
+
+              within_ranges <- vapply(seq_len(nc), function(i) {
+                  pairs <-  c(ranges[2*i - 1], ranges[2*i])
+                  between(query[[i]], pairs)
+              }, logical(nrow(query)))
+
+              idc <- which(rowSums(within_ranges, na.rm = FALSE) == nc)
```
[the remark this answers is not in the document]
Comment (philouail) [reply R/Spectra.R:2440]: No I think `vapply` is different from `sapply` and `lapply` by also the fact that it return a matrix/array that depends on the specified output (here a logical of length `now(query)`), it is also *always* simplified compared to the other apply function from what i understand. so ye, the code is a bit weird with `vapply` but the output is perfect as it is directly a matrix and also much much faster than using for loop (did a little benchmarking).

File: R/Spectra.R [hunk 2]
```diff
@@ -471,6 +471,9 @@ NULL
 #'   `rt[2]`. Returns the filtered `Spectra` (with spectra in their
 #'   original order).
 #'
+#' - `filterRanges`: allows filtering of the `Spectra` object
+#'    based on specified ranges for *any* values of `spectraVariables(object)`.
```
[the remark this answers is not in the document]
Comment (philouail) [reply R/Spectra.R:475]: True, I forgot about adding something to the vignette ! Maybe I can wait until we decide to implement or not the second function to show an example ? 

I also detailed more the example in the meantime.

**Comment by philouail** [comment]
Thanks @jorainer ! I have adapted the code to your suggestions.

File: R/Spectra.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (jorainer) [reply R/Spectra.R:2452]: Nice, wasn't aware of that :) - great, thanks for the explanation!

File: R/Spectra.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (jorainer) [reply R/Spectra.R]: Yes, that makes sense.

**Review by jorainer — APPROVED** [review]
Thanks Phili! This function makes a lot of sense to me!

**Comment by jorainer** [comment]
Thanks for the review @lgatto ! Regarding the name, if I understand you you would like to name the function to something like `filterSpectraDataRanges`? Or could you provide a possible example for the name?

I would not use `spectraData` in the name of the function - I would understand that I would filter/subset the `spectraData` (i.e. remove certain columns of `spectraData`) instead of filtering the `Spectra`.

To me `filterRanges` would make sense, because we want to filter the `Spectra` using values in one or more `spectraVariables` (hence the name of that parameter being `spectraVariables`). Also, for the other method we would use `filterValues`. And yes, the idea was in fact that the method should be similar to e.g. `filterPrecursorMzRange` or `filterPrecursorMzValue`, so the user would find/understand the method right away. In facto `filterPrecursorMzRange` would be the same as calling `filterRanges(, spectraVariable = "precursorMz", ...)`.

We also thought of a generic `filterSpectra` method and a `RangeFilter` parameter - that would be another possibility, but then we would add a new S4 object (`RangeFilter`) and hence increase the complexity of `Spectra`...

**Comment by lgatto** [comment]
OK for `filterRanges()`, you convinced me. Please make sure in the documentation that `filterPrecursorMzRange()` and  `filterRanges(, spectraVariable = "precursorMz", ...)` are equivalent.

**Comment by philouail** [comment]
Thanks @lgatto for the feedback ! I will adjust the documentation and the code accordingly. As well as adding the `filterValues()` function and a explanation in the vignette. I will try to make it as clear as possible how they can be used :) Putting it in draft for now while i prepare it !

File: R/Spectra.R [hunk 3]
```diff
@@ -2396,6 +2426,40 @@ setMethod("reset", "Spectra", function(object, ...) {
...
 
+
+#' @rdname Spectra
+#' @importFrom MsCoreUtils between
+#' @export
+setMethod("filterRanges", "Spectra",
+          function(object, spectraVariables, ranges, ...){
+              if (is.character(spectraVariables)){
+                  if(!all(spectraVariables %in% spectraVariables(object)))
+                      stop("'spectraVariables' need to correspond to colnames of",
```
[the remark this answers is not in the document]
Comment (jorainer) [reply R/Spectra.R:2437]: Maybe specifically mention that we mean the *parameter* spectraVariables would help? I would keep/use the name `spectraVariables` for the parameter and then in the error message say that `One or more values passed with parameter 'spectraVariables' are not available as spectra variables in object. Use the 'spectraVariables()' function to list possible values.`.

just a suggestion...

**Comment by philouail** [comment]
So here is everything updated. I moved the main code to the MsBackend.R  to have a similar implementation as the other filtering functions. 

Some things i am not 100% sure about: 

- location of unit tests : is test_Spectra.R fine ?
- Should `tolerance` be by default `Inf` ? In `mclosest()` it is but `filterPrecursorMzValue()` is set as 0. 
- the param `values` name ? there is already a `value` one, just want to be sure it's not confusing.

Also curiosity question but what is the different between `filterPrecrusorMz`, `filterPrecursorMzValue`, and `filterPrecursorMzRange`.

I will write the vignette now while i wait for feedback :) Thanks again for the help !

**Comment by jorainer** [comment]
Regarding your last question: `filterPrecursorMz` and `filterPrecursorMzValues` are the same (i.e. match/filter based on similarity of the precursor m/z). Both (should?) support to match against multiple *target* m/z values (e.g. `filterPrecursorMzValues(sps, mz = c(123.3, 234.2))` to keep spectra with a precursor m/z that matches any of the provided values). The `filterPrecursorMzRange` allows to define a broader range (i.e. precursor m/z within a specified range).

**Comment by jorainer** [comment]
For the default of `tolerance`. For `filterValues` I would use `0` as default. also for other methods (`filterPrecursorMzValues`) we have `0`. The `mclosest` (and `closest`) are a bit different than these filter functions: from a filter function I would expect to subset the data based on a (stringent) condition. From `closest` `mclosest` I would just like to get the *most similar* hits. The `tolerance` there is essentially a way to define that something is too different to be considered *close*. Hope this does make sense?

**Review by jorainer — CHANGES_REQUESTED** [review]
Looks good @philouail ! Thanks! I have some change requests - mostly related to documentation - and one suggestion to maybe add an additional parameter to both functions that allows to combine the multiple filters/conditions either with a logical and (the default) or a logical or.

File: R/Spectra.R [hunk 4]
```diff
@@ -471,6 +471,23 @@ NULL
 #'   `rt[2]`. Returns the filtered `Spectra` (with spectra in their
 #'   original order).
 #'
+#' - `filterRanges`: allows filtering of the `Spectra` object based on user
```
Comment (jorainer) [inline R/Spectra.R:474]: please update the help for `filterRanges` and `filterValues` with the updated versions suggested in `MsBackend.R` (except of the "default implementation" mention.

File: R/Spectra.R [hunk 5]
```diff
@@ -807,6 +824,10 @@ NULL
 #'     `c("mz", "value")` but any values returned by `peaksVariables(object)`
 #'     with `object` being the `Spectra` object are supported.
 #'
+#' @param condition For `filterRanges` and `filterValues` (default "all") can
```
Comment (jorainer) [inline R/Spectra.R:827]: also this, please update with the documentation in MsBackend.R

File: R/Spectra.R [hunk 6]
```diff
@@ -971,11 +1001,16 @@ NULL
...
 #'
-#' @param spectraVariables For `selectSpectraVariables`: `character` with the
+#' @param spectraVariables
+#' - For `selectSpectraVariables`: `character` with the
 #'     names of the spectra variables to which the backend should be subsetted.
-#'     For `addProcessing`: `character` with additional spectra variables that
+#' - For `addProcessing`: `character` with additional spectra variables that
 #'     should be passed along to the function defined with `FUN`. See function
 #'     description for details.
+#' - For `filterRanges` and `filterValues`: A `character` vector specifying the
```
Comment (jorainer) [inline R/Spectra.R:1010]: maybe also mention:

... vector specifying the column(s) from `spectraData(object)` on which to filter the data and that correspond to the ...

File: tests/testthat/test_Spectra.R [hunk 7]
```diff
@@ -1096,6 +1096,93 @@ test_that("filterRt,Spectra works", {
...
+test_that("filterRanges, Spectra works", {
+    filt_spectra <- filterRanges(sps_dia, spectraVariables = c("rtime",
+                                                               "precursorMz",
+                                                               "peaksCount"),
+                                 ranges = c(30, 350, 200,500, 350, 600))
+    # test does not accept logical
+    logical_test <- spectraVariables(sps_dia) %in% c("rtime", "precursorMz",
+                                                     "peaksCount")
+    expect_error(filterRanges(sps_dia, spectraVariables = logical_test,
+                              ranges = c(30, 350, 200,500, 350, 600)))
```
Comment (jorainer) [inline tests/testthat/test_Spectra.R:1108]: please for `expect_error` always also check for the error message - that the one you expect is thrown. In this case simply do `expect_error(... , "character")` (i.e. use one of the words from the error message that is thrown).

File: vignettes/Spectra.Rmd [hunk 8]
```diff
@@ -393,7 +393,17 @@ affecting the total number of spectra are:
   precursor charge(s).
 - `filterPrecursorScan`: retains (parent and children) scans of an acquisition
   number.
+- `filterRanges`: retains spectra within user defined `ranges` for one or more
+  available `spectraVariables` in object. if multiple ranges/spectra variables
```
Comment (jorainer) [inline vignettes/Spectra.Rmd:397]: Please use `. If multiple ...` and maybe also update with the suggested change for the documentation above for MsBackend.R

File: vignettes/Spectra.Rmd [hunk 9]
```diff
@@ -415,9 +425,48 @@ peaks data (m/z and intensity values) within each spectrum are:
 - `reduceSpectra`: filters individual spectra keeping only the largest peak for
   groups of peaks with similar m/z values.
 
-In the example below we select all spectra measured in the second mzML file and
-subsequently filter them to retain spectra measured between 175 and 189 seconds
-in the measurement run.
+In this example, we use the filterValues function to retain spectra with a base
```
Comment (jorainer) [inline vignettes/Spectra.Rmd:428]: please use the back ticks (\`) here around `filterValues`

File: vignettes/Spectra.Rmd [hunk 10]
```diff
@@ -415,9 +425,48 @@ peaks data (m/z and intensity values) within each spectrum are:
...
+peak m/z close to 100 (+/- 30 ppm) and a retention time around 230 (+/- 5 s).
+
+```{r}
+sps_sub <- filterValues(sps_sciex, spectraVariables = c("basePeakMZ", "rtime"),
```
Comment (jorainer) [inline vignettes/Spectra.Rmd:432]: Nice example with the base peak m/z - was not even aware that we have that spectra variable - and we definitely don't have a filter function for that particular variable.

Only, please always add a whitespace after the `,` to make the code a little more readable.

File: vignettes/Spectra.Rmd [hunk 11]
```diff
@@ -415,9 +425,48 @@ peaks data (m/z and intensity values) within each spectrum are:
...
+```{r}
+sps_ranges <- filterRanges(sps_sciex,
+                           spectraVariables = c("basePeakMZ","peaksCount",
+                                                "rtime"),
+                           ranges = c(123.09,124, 3500, 3520, 259, 260),
+                           condition = "all")
+length(sps_ranges)
+```
+
+Only one spectrum matches all the ranges. Another option for filterValues and
```
Comment (jorainer) [inline vignettes/Spectra.Rmd:451]: please always add backticks around actual function (or object) names as well as for R code, such as `condition = "any"`.

File: vignettes/Spectra.Rmd [hunk 12]
```diff
@@ -415,9 +425,48 @@ peaks data (m/z and intensity values) within each spectrum are:
...
+filterRanges is to use the parameter condition = "any", which retains spectra
+that match any one of the conditions instead of having to match all of them.
+Let's run the code oncwe again but change the condition parameter this time:
```
Comment (jorainer) [inline vignettes/Spectra.Rmd:454]: oncwe -> once

**Review by jorainer — CHANGES_REQUESTED** [review]
Really great contribution Phili! Thanks! 

I'm totally fine with the implementations, but have some suggestions to improve/fix the documentation.

File: R/MsBackend.R [hunk 13]
```diff
@@ -67,6 +67,10 @@
 #'     `peaksVariables(object)` and depends on what *peaks variables* the
 #'     backend provides.
 #'
+#' @param condition For `filterRanges` and `filterValues` (default "all") can
```
[the remark this answers is not in the document]
Comment (philouail) [reply R/MsBackend.R:70]: I agree with the naming of the param ! it makes much more sense

**Review by jorainer — APPROVED** [review]
Thanks Phili! All good from my side!

File: R/MsBackend.R [hunk 14]
```diff
@@ -136,21 +146,38 @@
 #' @param polarity For `filterPolarity`: `integer` specifying the polarity to
 #'     to subset `object`.
 #'
+#' @param ranges for `filterRanges`: A `numeric` vector of paired values (upper
```
[the remark this answers is not in the document]
Comment (philouail) [reply R/MsBackend.R:149]: So `ranges` length is not finite, it's lenght should be 2*spectraVariables if that helps to make sense. Essentially the values go 2 by 2 but ranges can be as long as the user wants. I can try to improve the overall definition of filterRanges to make that more clear maybe ? I'm also hoping that the examples and vignette helps ?

File: R/MsBackend.R [hunk 15]
```diff
@@ -121,6 +126,11 @@
 #' @param ppm For `filterPrecursorMzValues`: `numeric(1)` with the m/z-relative
 #'     maximal acceptable difference for a m/z to be considered matching. See
 #'     [closest()] for details.
+#'     For `filterValues`: `numeric` of any length allowing to define
```
[the remark this answers is not in the document]
Comment (philouail) [reply R/MsBackend.R:129]: Yes I can do that in the next push !

File: R/MsBackend.R (hunk 14 above)
[the remark this answers is not in the document]
Comment (lgatto) [reply R/MsBackend.R:149]: Yes, that's ok then, thanks.

**Review by lgatto — APPROVED** [review]

**Comment by lgatto** [comment]
Thank you @philouail for the great contribution!

File: R/MsBackend.R (hunk 15 above)
[the remark this answers is not in the document]
Comment (jorainer) [reply R/MsBackend.R:129]: yes, that makes sense - can you then please do that for all functions - so that we have everything consistent.

**Pull request merged**